File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleUserDetails.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.Immutable
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.Saver
import androidx.compose.runtime.saveable.listSaver

/** The fields the design labels "attached to every job", which is why every product collects them. */
@Immutable
data class UseSmileIDSampleUserDetails(
    val firstName: String = "",
    val lastName: String = "",
    val email: String = "",
    val phone: String = "",
) {
    /** The design's own rule: "First and last name are required." */
    val isComplete: Boolean get() = firstName.isNotBlank() && lastName.isNotBlank()

    /** Whether the form has collected what [requirement] still asks of it. */
    fun satisfies(requirement: UseSmileIDSampleUserDetailsRequirement): Boolean =
        (!requirement.firstName || firstName.isNotBlank()) &&
            (!requirement.lastName || lastName.isNotBlank()) &&
            (!requirement.contact || email.isNotBlank() || phone.isNotBlank())

    companion object {
        val Saver: Saver<MutableState<UseSmileIDSampleUserDetails>, Any> = listSaver(
            save = { listOf(it.value.firstName, it.value.lastName, it.value.email, it.value.phone) },
            restore = { mutableStateOf(UseSmileIDSampleUserDetails(it[0], it[1], it[2], it[3])) },
        )
    }
}

/**
 * What the form must still collect: the SDK's rule minus what the token binds. The static `required` flags
 * below cannot express either half of that.
 */
@Immutable
data class UseSmileIDSampleUserDetailsRequirement(
    val firstName: Boolean = true,
    val lastName: Boolean = true,
    val contact: Boolean = true,
) {
    /** Nothing left to ask, so the form has no reason to appear. */
    val isSatisfied: Boolean get() = !firstName && !lastName && !contact

    /** No relevant binding at all, which is the one case the SDK's own validator can still decide. */
    val bindsNothing: Boolean get() = firstName && lastName && contact

    /** Whether [field] is one the token already supplied, which is why it renders as provided. */
    fun supplies(field: UseSmileIDSampleUserField): Boolean = when (field) {
        UseSmileIDSampleUserField.FirstName -> !firstName
        UseSmileIDSampleUserField.LastName -> !lastName
        // Neither contact row is individually supplied: the rule is "one of", so a bound email leaves
        // phone askable and vice versa. Only the requirement itself lifts.
        UseSmileIDSampleUserField.Email, UseSmileIDSampleUserField.Phone -> false
    }

    /** A contact row stops saying "optional" the moment one of the two is actually required. */
    fun labelFor(field: UseSmileIDSampleUserField): String = when {
        !contact -> field.label
        field == UseSmileIDSampleUserField.Email -> "Email"
        field == UseSmileIDSampleUserField.Phone -> "Phone"
        else -> field.label
    }

    /** The sentence under the form, which has to name what is actually outstanding. */
    val prompt: String
        get() = buildList {
            if (firstName) add("first name")
            if (lastName) add("last name")
            if (contact) add("an email or phone number")
        }.let { outstanding ->
            when {
                outstanding.isEmpty() -> "Tap any field to edit."
                outstanding.size == 1 -> "${outstanding.single().replaceFirstChar(Char::titlecase)} is required."
                else -> "Required: ${outstanding.joinToString(", ")}."
            }
        }
}

/**
 * The requirement a token leaves behind. Mirrors the SDK's union rule field for field, and the unit
 * tests pin it to `FlowValidator.validateUserDetails`, which is the authority.
 */
fun UseSmileIDSampleTokenBindings?.userDetailsRequirement() = UseSmileIDSampleUserDetailsRequirement(
    firstName = this?.givenNames != true,
    lastName = this?.lastName != true,
    contact = !(this?.email == true || this?.phoneNumber == true),
)

/** Which user-details row changed, so the form reports one callback rather than four. */
enum class UseSmileIDSampleUserField(val id: String, val label: String, val placeholder: String, val required: Boolean) {
    FirstName("firstName", "First name", "Add first name", required = true),
    LastName("lastName", "Last name", "Add last name", required = true),
    Email("email", "Email (optional)", "name@company.com", required = false),
    Phone("phone", "Phone (optional)", "+254 700 000 000", required = false),
    ;

    fun read(details: UseSmileIDSampleUserDetails) = when (this) {
        FirstName -> details.firstName
        LastName -> details.lastName
        Email -> details.email
        Phone -> details.phone
    }

    fun write(details: UseSmileIDSampleUserDetails, value: String) = when (this) {
        FirstName -> details.copy(firstName = value)
        LastName -> details.copy(lastName = value)
        Email -> details.copy(email = value)
        Phone -> details.copy(phone = value)
    }
}

/** The ID-details form. ID type stays unselectable until a country is chosen, which the form is built around. */
@Immutable
data class UseSmileIDSampleIdDetails(
    val country: UseSmileIDSampleCountry? = null,
    val idType: UseSmileIDSampleIdType? = null,
    val idNumber: String = "",
) {
    val isComplete: Boolean get() = country != null && idType != null && idNumber.isNotBlank()
}

/** A country the sandbox supports, with the flag the picker leads each row with. */
enum class UseSmileIDSampleCountry(val code: String, val label: String, val flag: String) {
    Nigeria("NG", "Nigeria", "🇳🇬"),
    Kenya("KE", "Kenya", "🇰🇪"),
    Ghana("GH", "Ghana", "🇬🇭"),
    SouthAfrica("ZA", "South Africa", "🇿🇦"),
    Uganda("UG", "Uganda", "🇺🇬"),
    Tanzania("TZ", "Tanzania", "🇹🇿"),
    Rwanda("RW", "Rwanda", "🇷🇼"),
}

/** ID types, filtered by country, which is why the trigger is disabled until one is chosen. */
enum class UseSmileIDSampleIdType(val id: String, val label: String, val countries: Set<UseSmileIDSampleCountry>) {
    NationalId("NATIONAL_ID", "National ID", UseSmileIDSampleCountry.entries.toSet()),
    Passport("PASSPORT", "Passport", UseSmileIDSampleCountry.entries.toSet()),
    DriversLicense("DRIVERS_LICENSE", "Driver's licence", setOf(UseSmileIDSampleCountry.Nigeria, UseSmileIDSampleCountry.Kenya, UseSmileIDSampleCountry.SouthAfrica)),
    VoterId("VOTER_ID", "Voter ID", setOf(UseSmileIDSampleCountry.Nigeria, UseSmileIDSampleCountry.Ghana)),
    ;

    companion object {
        fun of(country: UseSmileIDSampleCountry?) = entries.filter { country != null && country in it.countries }
    }
}

File: ios/SampleUI/Sources/SampleUI/State/UseSmileIDSampleUserDetails.swift
import Foundation

/// The fields the design labels "attached to every job", which is why every product collects them.
public struct UseSmileIDSampleUserDetails: Equatable, Sendable {
  public var firstName: String
  public var lastName: String
  public var email: String
  public var phone: String

  public init(firstName: String = "", lastName: String = "", email: String = "", phone: String = "") {
    self.firstName = firstName
    self.lastName = lastName
    self.email = email
    self.phone = phone
  }

  /// The design's own rule: "First and last name are required."
  public var isComplete: Bool {
    !firstName.isBlank && !lastName.isBlank
  }

  /// Whether the form has collected what `requirement` still asks of it.
  public func satisfies(_ requirement: UseSmileIDSampleUserDetailsRequirement) -> Bool {
    (!requirement.firstName || !firstName.isBlank)
      && (!requirement.lastName || !lastName.isBlank)
      && (!requirement.contact || !email.isBlank || !phone.isBlank)
  }
}

/// What the form must still collect: the SDK's rule minus what the token binds.
public struct UseSmileIDSampleUserDetailsRequirement: Equatable, Sendable {
  public var firstName: Bool
  public var lastName: Bool
  public var contact: Bool

  public init(firstName: Bool = true, lastName: Bool = true, contact: Bool = true) {
    self.firstName = firstName
    self.lastName = lastName
    self.contact = contact
  }

  /// Nothing left to ask, so the form has no reason to appear.
  public var isSatisfied: Bool {
    !firstName && !lastName && !contact
  }

  /// No relevant binding at all, the one case the SDK's own validator can still decide.
  public var bindsNothing: Bool {
    firstName && lastName && contact
  }

  /// Whether the token already supplied `field`, which is why it renders as provided.
  public func supplies(_ field: UseSmileIDSampleUserField) -> Bool {
    switch field {
    case .firstName: !firstName
    case .lastName: !lastName
    // The rule is "one of", so a bound email leaves phone askable. Only the requirement lifts.
    case .email, .phone: false
    }
  }

  /// A contact row stops saying "optional" the moment one of the two is actually required.
  public func label(for field: UseSmileIDSampleUserField) -> String {
    guard contact else { return field.label }
    switch field {
    case .email: return "Email"
    case .phone: return "Phone"
    default: return field.label
    }
  }

  /// The sentence under the form, which has to name what is actually outstanding.
  public var prompt: String {
    var outstanding: [String] = []
    if firstName {
      outstanding.append("first name")
    }
    if lastName {
      outstanding.append("last name")
    }
    if contact {
      outstanding.append("an email or phone number")
    }
    switch outstanding.count {
    case 0: return "Tap any field to edit."
    case 1: return outstanding[0].prefix(1).uppercased() + outstanding[0].dropFirst() + " is required."
    default: return "Required: " + outstanding.joined(separator: ", ") + "."
    }
  }
}

public extension UseSmileIDSampleUserDetailsRequirement {
  /// What a token leaves the form to collect, mirroring the SDK's union rule: both names plus one contact field.
  init(bindings: UseSmileIDSampleTokenBindings?) {
    self.init(
      firstName: bindings?.givenNames != true,
      lastName: bindings?.lastName != true,
      contact: !(bindings?.email == true || bindings?.phoneNumber == true)
    )
  }
}

/// Which user-details row changed, so the form reports one callback rather than four.
public enum UseSmileIDSampleUserField: String, CaseIterable, Sendable {
  case firstName
  case lastName
  case email
  case phone

  public var id: String {
    rawValue
  }

  public var label: String {
    switch self {
    case .firstName: "First name"
    case .lastName: "Last name"
    case .email: "Email (optional)"
    case .phone: "Phone (optional)"
    }
  }

  public var placeholder: String {
    switch self {
    case .firstName: "Add first name"
    case .lastName: "Add last name"
    case .email: "name@company.com"
    case .phone: "+254 700 000 000"
    }
  }

  public var required: Bool {
    self == .firstName || self == .lastName
  }

  public func read(_ details: UseSmileIDSampleUserDetails) -> String {
    switch self {
    case .firstName: details.firstName
    case .lastName: details.lastName
    case .email: details.email
    case .phone: details.phone
    }
  }

  public func write(_ details: UseSmileIDSampleUserDetails, _ value: String) -> UseSmileIDSampleUserDetails {
    var next = details
    switch self {
    case .firstName: next.firstName = value
    case .lastName: next.lastName = value
    case .email: next.email = value
    case .phone: next.phone = value
    }
    return next
  }
}

extension String {
  /// Whitespace-only counts as missing: a space is not a first name.
  var isBlank: Bool {
    trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
  }
}
